diff_drive_controller/test/test_diff_drive_controller.cpp

```diff
@@ -0,0 +1,325 @@
+
+  void SetUp() override
+  {
+    test_robot = std::make_shared<test_robot_hardware::TestRobotHardware>();
+    test_robot->init();
+    left_wheel_names =
+    {{test_robot->joint_name1, test_robot->joint_name2, test_robot->joint_name3}};
+    right_wheel_names =
+    {{test_robot->joint_name1, test_robot->joint_name2, test_robot->joint_name3}};
+    op_mode = {{test_robot->write_op_handle_name1}};
+
+    pub_node = std::make_shared<rclcpp::Node>("velocity_publisher");
+    velocity_publisher = pub_node->create_publisher<geometry_msgs::msg::TwistStamped>(
+      controller_name + "/cmd_vel",
+      rclcpp::SystemDefaultsQoS());
+  }
+
+  static void TearDownTestCase()
+  {
+    rclcpp::shutdown();
+  }
+
+  /// Publish velocity msgs
+  /**
+   *  linear - magnitude of the linear command in the geometry_msgs::twist message
+   *  angular - the magnitude of the angular command in geometry_msgs::twist message
+   */
+  void publish(double linear, double angular)
+  {
+    int wait_count = 0;
+    auto topic = velocity_publisher->get_topic_name();
+    while (pub_node->count_subscribers(topic) == 0) {
+      if (wait_count >= 5) {
+        auto error_msg = std::string("publishing to ") + topic + " but no node subscribes to it";
+        throw std::runtime_error(error_msg);
+      }
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+      ++wait_count;
+    }
+
+    geometry_msgs::msg::TwistStamped velocity_message;
+    velocity_message.header.stamp = pub_node->get_clock()->now();
+    velocity_message.twist.linear.x = linear;
+    velocity_message.twist.angular.z = angular;
+    velocity_publisher->publish(velocity_message);
+  }
+
+  /// \brief wait for the subscriber and publisher to completely setup
+  void waitForSetup()
+  {
+    constexpr std::chrono::seconds TIMEOUT{2};
+    auto clock = pub_node->get_clock();
+    auto start = clock->now();
+    while (velocity_publisher->get_subscription_count() <= 0) {
+      if ((clock->now() - start) > TIMEOUT) {
+        FAIL();
+      }
+      rclcpp::spin_some(pub_node);
+    }
+  }
+
+  void setup_controller(
+    TestableDiffDriveController & controller,
+    rclcpp::Executor & executor)
+  {
+    auto diff_drive_lifecycle_node = controller.get_lifecycle_node();
+    executor.add_node(diff_drive_lifecycle_node->get_node_base_interface());
+    auto state = diff_drive_lifecycle_node->configure();
+    ASSERT_EQ(State::PRIMARY_STATE_INACTIVE, state.id());
+
+    state = diff_drive_lifecycle_node->activate();
+    ASSERT_EQ(State::PRIMARY_STATE_ACTIVE, state.id());
+
+    waitForSetup();
+  }
+
+  std::string controller_name = "test_diff_drive_controller";
+
+  std::shared_ptr<test_robot_hardware::TestRobotHardware> test_robot;
+  std::vector<std::string> left_wheel_names;
+  std::vector<std::string> right_wheel_names;
+  std::vector<std::string> op_mode;
+
+  rclcpp::Node::SharedPtr pub_node;
+  rclcpp::Publisher<geometry_msgs::msg::TwistStamped>::SharedPtr velocity_publisher;
```

Comment:
Add a trailing underscore for member variables.

Reply:
roger roger